devices/flash-imxrt/flashdrv.c

```diff
@@ -198,81 +142,133 @@ s32 flashdrv_bufferedPagesWrite(flash_context_t *ctx, u32 offset, const char *bu
 
 		/* Save filled buffer */
 		if (ctx->counter >= ctx->properties.sector_size)
-			flashdrv_sync(ctx);
+			flashdrv_syncCtx(ctx);
 	}
 
 	return size;
 }
 
 
-int flashdrv_chipErase(flash_context_t *ctx)
+static int flashdrv_defineFlexSPI(flash_context_t *ctx)
 {
-	return flexspi_norFlashEraseAll(ctx->instance, &ctx->config);
+	switch (ctx->address) {
+		case FLASH_FLEXSPI1:
+			ctx->instance = FLASH_FLEXSPI1_INSTANCE;
+			ctx->maxSize = FLASH_SIZE_FLEXSPI1;
+			ctx->option.option0 = FLASH_FLEXSPI1_QSPI_FREQ;
+			ctx->option.option1 = 0;
+			break;
+
+		case FLASH_FLEXSPI2:
+			ctx->instance = FLASH_FLEXSPI2_INSTANCE;
+			ctx->maxSize = FLASH_SIZE_FLEXSPI2;
+			ctx->option.option0 = FLASH_FLEXSPI2_QSPI_FREQ;
+			ctx->option.option1 = 0;
+			break;
+
+		default:
+			return ERR_ARG;
+	}
+
+	return ERR_NONE;
 }
 
 
-int flashdrv_sectorErase(flash_context_t *ctx, u32 offset)
+/* Device interafce */
+
+static ssize_t flashdrv_read(unsigned int dn, addr_t offs, u8 *buff, unsigned int len)
 {
-	if (offset % ctx->properties.sector_size)
-		return -1;
+	ssize_t res;
+
+	if (dn >= FLASH_NO)
+		return ERR_ARG;
+
+	if ((res = flashdrv_readData(&flashdrv_common.ctx[dn], offs, (char *)buff, len)) < 0)
+		return res;
 
-	return flexspi_norFlashErase(ctx->instance, &ctx->config, offset, ctx->properties.sector_size);
+	return res;
 }
 
 
-void flashdrv_sync(flash_context_t *ctx)
+static ssize_t flashdrv_write(unsigned int dn, addr_t offs, const u8 *buff, unsigned int len)
 {
-	int i;
-	u32 dstAddr;
-	const u32 *src;
-	const u32 pagesNumber = ctx->properties.sector_size / ctx->properties.page_size;
+	size_t size;
+	addr_t bOffs = 0;
+	flash_context_t *ctx;
 
-	if (ctx->counter == 0)
-		return;
+	size = len;
+	if (!len)
+		return ERR_NONE;
 
-	for (i = 0; i < pagesNumber; ++i) {
-		dstAddr = ctx->sectorID * ctx->properties.sector_size + i * ctx->properties.page_size;
-		src = (const u32 *)(ctx->buff + i * ctx->properties.page_size);
+	if (dn >= FLASH_NO || offs % 0x100)
+		return ERR_ARG;
 
-		flexspi_norFlashPageProgram(ctx->instance, &ctx->config, dstAddr, src);
+	ctx = &flashdrv_common.ctx[dn];
+
+	while (size) {
+		if (size < 0x100) {
```

Comment:
Maybe page size should be defined in one place? It is popular size for sure, but there is no guarantee that no flash memory will have different page size